source/extensions/filters/network/proxy_protocol/proxy_protocol.cc

```diff
@@ -0,0 +1,185 @@
+#include "extensions/filters/network/proxy_protocol/proxy_protocol.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/network/address_impl.h"
+#include "common/tcp_proxy/tcp_proxy.h"
+
+#include "extensions/filters/common/proxy_protocol/proxy_protocol_header.h"
+
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V1_AF_INET;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V1_AF_INET6;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V1_SIGNATURE;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_ADDR_LEN_INET;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_ADDR_LEN_INET6;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_AF_INET;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_AF_INET6;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_ONBEHALF_OF;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_SIGNATURE;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_SIGNATURE_LEN;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_TRANSPORT_STREAM;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_VERSION;
+using envoy::extensions::filters::network::proxy_protocol::v3::ProxyProtocol_Version;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ProxyProtocol {
+
+Filter::Filter(const envoy::extensions::filters::network::proxy_protocol::v3::ProxyProtocol& config)
+    : config_(config) {}
+
+Network::FilterStatus Filter::onWrite(Buffer::Instance&, bool) {
+  if (!sent_header_) {
+    injectHeader();
+    sent_header_ = true;
+  }
+  return Network::FilterStatus::Continue;
+}
+
+void Filter::injectHeader() {
+  if (config_.version() == ProxyProtocol_Version::ProxyProtocol_Version_V1) {
+    injectHeaderV1();
+  } else if (config_.version() == ProxyProtocol_Version::ProxyProtocol_Version_V2) {
+    injectHeaderV2();
+  }
+}
+
+void Filter::injectHeaderV1() {
+  std::ostringstream stream;
+  stream << PROXY_PROTO_V1_SIGNATURE;
+  // Default to local addresses
+  std::string src_address = write_callbacks_->connection().localAddress()->ip()->addressAsString();
+  std::string dst_address = write_callbacks_->connection().remoteAddress()->ip()->addressAsString();
+  auto src_port = write_callbacks_->connection().localAddress()->ip()->port();
+  auto dstPort = write_callbacks_->connection().remoteAddress()->ip()->port();
+  auto ip_version = write_callbacks_->connection().localAddress()->ip()->version();
+
+  if (write_callbacks_->connection().streamInfo().filterState()->hasData<TcpProxy::DownstreamAddrs>(
```

Comment:
When would we use this?
I had assumed we'd want it for if we were replacing an incoming proxy protocol header, but it looks to me like listener/proxy_protocol.cc actually sets the remote and local address so anyone looking at the socket sees the addresses from the header.  Do you have a use case where we'd want to do something different than that?

Reply:
The use case is clients talking to Envoy that don't send the protocol, and Envoy generating the PROXY protocol header on behalf of those clients to send to backends that expect the header. Essentially mirrors the `send-proxy` setting in HAProxy, making it possible to swap in Envoy for HAProxy in cases like this.